geopandas/tools/overlay.py

```diff
@@ -76,15 +69,19 @@ def _overlay_difference(df1, df2):
     Overlay Difference operation used in overlay function
     """
     # Spatial Index to create intersections
-    spatial_index = df2.sindex
-    bbox = df1.geometry.apply(lambda x: x.bounds)
-    sidx = bbox.apply(lambda x: list(spatial_index.intersection(x)))
+    df1_idx, df2_idx = df2.sindex.query_bulk(
+        df1.geometry, predicate="intersects", sort=True
+    )
+    df2_geoms = (
+        df2.geometry.values[df2_idx[df1_idx == idx]]
+        if idx in np.unique(df1_idx)
+        else []
+        for idx in range(df1.geometry.size)
+    )
     # Create differences
     new_g = []
-    for geom, neighbours in zip(df1.geometry, sidx):
-        new = reduce(
-            lambda x, y: x.difference(y), [geom] + list(df2.geometry.iloc[neighbours])
-        )
+    for geom, neighbours in zip(df1.geometry, df2_geoms):
+        new = reduce(lambda x, y: x.difference(y), [geom] + list(neighbours))
         new_g.append(new)
     differences = GeoSeries(new_g, index=df1.index, crs=df1.crs)
```

Comment:
If we are able to only loop over the intersecting geometries in the loop above, can we do `differences = df1.geometry.copy()` and then assign `differences[df1_idx] = new_g` or something like that?